### PR TITLE
Fix wrong dubbo trace caused by using rpcContext.isProviderSide()

### LINKED ISSUE #12928: The Dubbo plugin may have a broken trace problem
### Describe the bug

About Dubbo instrumentation, from our customers, we found it will return expected result:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/02efad2ab9e6ae23245204f90e7627c371a9e1a3/instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/TracingFilter.java#L43
In provider side, it might returns false and causes broken trace with upstream client's trace.
But I can't reproduce it in my local environment, it might causes by several factors. 
From our customers' use cases, it will happens in Dubbo 2.7.10. I also talked with maintainer @AlbumenJ from Dubbo community. He expressed we shouldn't use it continuously and he doesn't recommend using it to check the source.
I also noticed there are some related issues in Dubbo community. For detail, please refer to https://github.com/apache/dubbo/issues?q=is%3Aissue+rpccontext+isconsumerside+is%3Aclosed

### Steps to reproduce

But I can't reproduce it in my local environment, it might causes by several factors. I found it by our customers.

### Expected behavior

The Dubbo instrumentation should produce correct trace.

### Actual behavior

In some scenarios, it will produces broken trace due to wrong return of `rpcContext.isProviderSide()`

### Javaagent or library instrumentation version

2.10.0

### Environment

**JDK**: JDK8
**OS**: CentOS

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by steverao**
Resolves #12928, I split the `OpenTelemetryFilter` into `OpenTelemetryClientFilter` and `OpenTelemetryServerFilter` and avoid using `rpcContext.isProviderSide()`.

### CODE REVIEW

File: instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java [hunk 1]
```diff
@@ -13,9 +13,18 @@
...
+  public static DubboTelemetry createClient(OpenTelemetry openTelemetry) {
+    return builder(openTelemetry).buildClient();
+  }
+
+  /**
+   * Returns a new server {@link DubboTelemetry} configured with the given {@link OpenTelemetry}.
+   */
+  public static DubboTelemetry createServer(OpenTelemetry openTelemetry) {
+    return builder(openTelemetry).buildServer();
   }
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java:28]: This does not align with existing instrumentations that instrument both client and server, such as https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/grpc-1.6/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_6, where `TelemetryBuilder` builds a single `Telemetry` that can be used to produce either client or server interceptor.

Comment (steverao) [reply instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java]: > This does not align with existing instrumentations that instrument both client and server, such as https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/grpc-1.6/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_6, where `TelemetryBuilder` builds a single `Telemetry` that can be used to produce either client or server interceptor.

Thanks, I modified it, Please have a look again!

File: instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryClientFilter.java [hunk 2]
```diff
@@ -16,20 +16,20 @@
 import org.apache.dubbo.rpc.Invoker;
 import org.apache.dubbo.rpc.Result;
 
-@Activate(group = {"consumer", "provider"})
-public class OpenTelemetryFilter implements Filter {
+@Activate(group = {"consumer"})
+public class OpenTelemetryClientFilter implements Filter {
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryClientFilter.java:20]: could be final

File: instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryServerFilter.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import io.opentelemetry.instrumentation.api.incubator.semconv.net.PeerServiceAttributesExtractor;
+import io.opentelemetry.javaagent.bootstrap.internal.AgentCommonConfig;
+import org.apache.dubbo.common.extension.Activate;
+import org.apache.dubbo.rpc.Filter;
+import org.apache.dubbo.rpc.Invocation;
+import org.apache.dubbo.rpc.Invoker;
+import org.apache.dubbo.rpc.Result;
+
+@Activate(group = {"provider"})
+public class OpenTelemetryServerFilter implements Filter {
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryServerFilter.java:20]: could be final

File: instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryServerFilter.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  private final Filter delegate;
+
+  public OpenTelemetryServerFilter() {
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryServerFilter.java:24]: server and client filter are almost identical, did you consider sharing some code between them?

File: instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java [hunk 5]
```diff
@@ -26,6 +26,7 @@ public static DubboTelemetryBuilder builder(OpenTelemetry openTelemetry) {
...
 
   private final Instrumenter<DubboRequest, Result> serverInstrumenter;
+
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java:29]:
```suggestion

```

File: instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java [hunk 6]
```diff
@@ -35,8 +36,13 @@ public static DubboTelemetryBuilder builder(OpenTelemetry openTelemetry) {
     this.clientInstrumenter = clientInstrumenter;
   }
 
-  /** Returns a new Dubbo {@link Filter} that traces Dubbo RPC invocations. */
-  public Filter newFilter() {
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java:39]: Usually we mark the old method as deprecated, keep it around for at least one release, and then remove.

Comment (steverao) [reply instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java:39]: > Usually we mark the old method as deprecated, keep it around for at least one release, and then remove.

I find it often appears in some situations where a change from experimental to stable. This is a bug here, I think we shouldn't allow users to use the method again. 
BTW, if it's necessary here, I found `TracingFilter` might need to keep the original `invoke` implementation again. It's a bit strange.

File: instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryServerFilter.java (hunk 4 above)
Comment (steverao) [reply instrumentation/apache-dubbo-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachedubbo/v2_7/OpenTelemetryServerFilter.java:24]: > server and client filter are almost identical, did you consider sharing some code between them?

Make sense, compared with other filters in library. It added extra attribute and causes a bit redundant here. If we extract a basic filter that contains them. The structure looks a bit complex. Or do you have any suggestion here?

File: instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java (hunk 6 above)
Comment (trask) [reply instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java:39]: one of the reasons we deprecate first before removing (when possible) is that we use the `@deprecated` Javadoc to tell users which method to use instead. you can remove the method in the following minor release

Comment (steverao) [reply instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTelemetry.java:39]: Thanks for your reminder and explanation, I have adjusted it and please have a look after holiday.

File: instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/TracingFilter.java [hunk 7]
```diff
@@ -40,9 +40,13 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) {
       return invoker.invoke(invocation);
     }
 
-    boolean isServer = rpcContext.isProviderSide();
-    Instrumenter<DubboRequest, Result> instrumenter =
-        isServer ? serverInstrumenter : clientInstrumenter;
+    Instrumenter<DubboRequest, Result> instrumenter;
+    if (clientInstrumenter != null && serverInstrumenter != null) {
```
Comment (laurit) [inline instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/TracingFilter.java:44]: imo the logic here is hard to follow, need to know when which of the instrumenters is set. I'd use something like https://github.com/steverao/opentelemetry-java-instrumentation/pull/7

Comment (steverao) [reply instrumentation/apache-dubbo-2.7/library-autoconfigure/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/TracingFilter.java]: > imo the logic here is hard to follow, need to know when which of the instrumenters is set. I'd use something like [steverao#7](https://github.com/steverao/opentelemetry-java-instrumentation/pull/7)

Thanks!

**Review by laurit — APPROVED** [review]

**Review by AlbumenJ — APPROVED** [review]
